### PR TITLE
Adding Unit test on assembly module, submodule download.py

### PR DESCRIPTION
**Pull request by ens-LCampbell**
In this PR:

# Moodule code changes:
- Small refactor to download.py submodule to facilitate easier unit test development. 
- Isolate FTP connection establishment to separate function. New function now performs the checking of accession for sub_dir path specification. 

# Unit test code:
- Added test functions for all functions 
- Includes small numbers of test data files with small footprint. 

- Included black reformatting to both module and unit test

` coverage run -m pytest && coverage report -m --rcfile=../../../../pyproject.toml`
`Name                                                  Stmts   Miss  Cover   Missing
src/python/ensembl/io/genomio/assembly/download.py       162      4    98%   233, 240, 248, 355`

### CODE REVIEW

**Review by MatBarba — APPROVED** [review]
Nicely done

File: src/python/ensembl/io/genomio/assembly/download.py [hunk 1]
```diff
@@ -102,47 +152,59 @@ def get_checksums(checksum_path: Path) -> Dict[str, str]:
...
+                f"Could not find accession '{accession}' from ftp {ftp_dir} in open FTP connection"
+            )
 
 
 def _download_file(
-    ftp_conn: FTP, ftp_file: str, md5_sums: Dict[str, str], dl_dir: Path, max_redo: int = 0
+    ftp_connection: FTP, ftp_file: str, md5_sums: Dict[str, str], dl_dir: Path, max_redo: int = 0
 ) -> None:
-    """TODO"""
+    """Download individual files from FTP server.
```
Comment (JAlvarezJarreta) [inline src/python/ensembl/io/genomio/assembly/download.py:196]:
```suggestion
    """Downloads individual files from FTP server.
```
Just a suggestion, as it seems the standard is to have descriptions start with a verb in third person :man_shrugging:

File: src/python/ensembl/io/genomio/assembly/download.py [hunk 2]
```diff
@@ -102,47 +152,59 @@ def get_checksums(checksum_path: Path) -> Dict[str, str]:
...
+
+    Args:
+        ftp_connection: Established connection FTP object.
+        ftp_file: Name of ftp file to download.
+        md5_sums: Dict of key value pairs filename<=>md5_checksums.
```
Comment (JAlvarezJarreta) [inline src/python/ensembl/io/genomio/assembly/download.py:201]:
```suggestion
        md5_sums: Dictionary of key value pairs filename - md5_checksums.
```

File: src/python/ensembl/io/genomio/assembly/download.py [hunk 3]
```diff
@@ -102,47 +152,59 @@ def get_checksums(checksum_path: Path) -> Dict[str, str]:
...
+        dl_dir: Path to downloaded FTP files.
+        max_redo: Maximum number of connection retry attempts.
+
+    Returns:
+        None
```
Comment (JAlvarezJarreta) [inline src/python/ensembl/io/genomio/assembly/download.py:206]:
```suggestion
```

File: src/python/ensembl/io/genomio/assembly/download.py [hunk 4]
```diff
@@ -211,7 +275,14 @@ def get_files_selection(dl_dir: Path) -> Dict[str, str]:
...
 def get_root_name(dl_dir: Path) -> str:
-    """Get root name for assembly files, using the report file as base"""
+    """Get root name for assembly files, using the report file as base
+
+    Args:
+        dl_dir: Path location of downloaded FTP files.
+
+    Returns:
+        Str: Shared download files basename prefix, obtained from asm report file.
+    """
```
Comment (JAlvarezJarreta) [inline src/python/ensembl/io/genomio/assembly/download.py:285]:
```suggestion
    """Returns the root name, i.e. shared files basename prefix, using the assembly report file as base.

    Args:
        dl_dir: Path location of downloaded FTP files.
    """
```

File: src/python/ensembl/io/genomio/assembly/download.py [hunk 5]
```diff
@@ -227,30 +298,26 @@ def retrieve_assembly_data(
     max_increment: int = 0,
     max_redo: int = 3,
 ) -> None:
-    """TODO
+    """Establish FTP connection and download a predefined subset of assembly data files from either
+    INSDC or RefSeq. For subset of files see Dict: '_FILE_ENDS'
```
Comment (JAlvarezJarreta) [inline src/python/ensembl/io/genomio/assembly/download.py:302]:
```suggestion
    """Establishes an FTP connection and downloads a predefined subset of assembly data files from either
    INSDC or RefSeq.
```

File: src/python/ensembl/io/genomio/assembly/download.py [hunk 6]
```diff
@@ -227,30 +298,26 @@ def retrieve_assembly_data(
...
 
     Args:
-        accession: Genome Assembly accession
-        download_dir: Path to directory used to store retrieved
-        max_increment: If you want to allow assembly versions
-        max_redo: Set max number of times to retry downloading a file
+        accession: Genome Assembly accession.
+        download_dir: Path to download FTP files.
+        max_increment: If you want to allow assembly versions.
+        max_redo: Maximum ftp connection retry attempts.
```
Comment (JAlvarezJarreta) [inline src/python/ensembl/io/genomio/assembly/download.py:308]:
```suggestion
        accession: Genome assembly accession.
        download_dir: Path to where to download FTP files.
        max_increment: If you want to allow assembly versions.
        max_redo: Maximum FTP connection retry attempts.
```

File: src/python/ensembl/io/genomio/assembly/download.py [hunk 7]
```diff
@@ -227,30 +298,26 @@ def retrieve_assembly_data(
...
 
+    Returns:
+        None
```
Comment (JAlvarezJarreta) [inline src/python/ensembl/io/genomio/assembly/download.py:311]:
```suggestion
    Raises:
        FileDownloadError: If no files are downloaded or if any does not match its MD5 checksum.
```

File: src/python/ensembl/io/genomio/assembly/download.py [hunk 8]
```diff
@@ -193,10 +253,14 @@ def _download_file(
...
-    The file_path is relative to the download dir
-    Current names are defined in _FILE_ENDS
+    Among all downloaded files only keep a subset for which we use a controlled name.
+    Current names are defined in dict: '_FILE_ENDS'
+
+    Args:
+        dl_dir: Local path to downloaded FTP files. The file_path is relative to the download dir.
+
+    Returns:
+        Dict: Subset target files. File name end (key)<=>file name(value)
```
Comment (JAlvarezJarreta) [inline src/python/ensembl/io/genomio/assembly/download.py:263]:
```suggestion
    """Returns a dictionary with the relevant downloaded files classified.

    Args:
        dl_dir: Local path to downloaded FTP files.

    Returns:
        Dictionary of file type (e.g.`"report"`) as keys and the relative file path (from `dl_dir`) as values.

    Raises:
        FileDownloadError: If `dl_dir` tree does not include a file named `*_assembly_report.txt`.
```
However, `FileDownloadError` should not be raised in this function, but in `get_root_name()` (based on the description and functionality of both methods).

Comment (JAlvarezJarreta) [reply src/python/ensembl/io/genomio/assembly/download.py:263]: Note: watch out that I have included the `"""` in my suggestion when they are in a separate line in the original code.

File: src/python/tests/assembly/test_download.py [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+    $ pytest ./assembly/test_download.py
+
+"""
+
+import filecmp
+import logging
+from pathlib import Path
+from unittest.mock import Mock, patch, MagicMock
+from contextlib import nullcontext as does_not_raise
+from typing import Callable, ContextManager
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:30]:
```suggestion
from contextlib import nullcontext as does_not_raise
import filecmp
import logging
from pathlib import Path
from typing import Callable, ContextManager, Optional
from unittest.mock import Mock, patch, MagicMock
```

File: src/python/tests/assembly/test_download.py [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+    ftp_url: str,
+    accession: str,
+    expectation: ContextManager,
+):
+    """Tests the FTPConnection method 'establish_ftp()'
+    Args:
+        mock_ftp: Mock FTP Object
+        ftp_url: FTP url
+        sub_dir: Sub directory path
+        expectation: Context manager expected raise exception
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:74]:
```suggestion

    Args:
        mock_ftp: Mock FTP object.
        ftp_url: FTP URL.
        sub_dir: Subdirectory path.
        expectation: Context manager expected raise exception.
```

File: src/python/tests/assembly/test_download.py [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+            Path("malformed_md5_checksums.txt"),
+            None,
+            pytest.raises(ValueError),
+            id="Bad md5 file - tab separated",
+        ),
+        pytest.param(Path("dir_name"), None, does_not_raise(), id="Dir passed instead of file"),
+    ],
+)
+def test_checksums(data_dir: Path, checksum_file: Path, checksum: str, expectation: ContextManager) -> None:
+    """Tests the 'download.get_checksums() function
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:108]:
```suggestion
    """Tests the `download.get_checksums()` function.
```

File: src/python/tests/assembly/test_download.py [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+        pytest.param(
+            Path("malformed_md5_checksums.txt"),
+            None,
+            pytest.raises(ValueError),
+            id="Bad md5 file - tab separated",
+        ),
+        pytest.param(Path("dir_name"), None, does_not_raise(), id="Dir passed instead of file"),
+    ],
+)
+def test_checksums(data_dir: Path, checksum_file: Path, checksum: str, expectation: ContextManager) -> None:
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:107]:
```suggestion
def test_checksums(
    data_dir: Path, checksum_file: Optional[Path], checksum: str, expectation: ContextManager
) -> None:
```

File: src/python/tests/assembly/test_download.py [hunk 13]
```diff
@@ -0,0 +1,452 @@
...
+        pytest.param("md5checksums.txt", None, True, id="Normal case"),
+        pytest.param("wrong_md5_checksums.txt", None, False, id="Incorrect md5 checksum"),
+        pytest.param(None, None, True, id="No md5file specified, resort to default"),
+        pytest.param(None, Path("*"), False, id="Incompatible os path '*'"),
+        pytest.param("missingfile_md5.txt", None, False, id="md5 checksum with ref of missing file"),
+    ],
+)
+@pytest.mark.dependency(depends=["test_checksums"])
+def test_md5_files(data_dir: Path, md5_file: str, md5_path: Path, checksum_bool: bool) -> None:
+    """Tests the md5_files() function"""
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:136]: Missing `Args` part.

File: src/python/tests/assembly/test_download.py [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+
+    mock_ftp.retrbinary.side_effect = mock_retr_binary
+
+    with expectation:
+        _download_file(mock_ftp, ftp_file, md5_sums, tmp_dir)
+        assert filecmp.cmp(data_file, retr_file)
+
+
+#################
+@pytest.mark.dependency(name="test_download_all_files")
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:220]:
```suggestion
@pytest.mark.dependency(name="test_download_all_files", depends=["test_download_single_file"])
```

File: src/python/tests/assembly/test_download.py [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+            "GCA_017607445.1",
+            "GCA_544706710.1",
+            "md5checksums_ftp_source.txt",
+            does_not_raise(),
+            2,
+            id="Error case: GCA doesn't match ftp_dir path",
+        ),
+    ],
+)
+@pytest.mark.dependency(depends=["test_download_single_file"])
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:244]:
```suggestion
```

File: src/python/tests/assembly/test_download.py [hunk 16]
```diff
@@ -0,0 +1,452 @@
...
+
+    mock_ftp.retrbinary.side_effect = mock_retr_binary
+
+    with exception:
+        connected_ftp = establish_ftp(mock_ftp, ftp_url, ftp_accession)
+        mock_ftp.connect.assert_called()
+        mock_ftp.login.assert_called()
+        mock_ftp.cwd.assert_called_with("genomes/all/GCA/017/607/445")
+
+    with exception:
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:297]:
```suggestion
```
You are already within the case context block above.

File: src/python/tests/assembly/test_download.py [hunk 17]
```diff
@@ -0,0 +1,452 @@
...
+            {},
+            pytest.raises(FileDownloadError),
+            id="Error case, data dir not provided",
+        ),
+    ],
+)
+def test_get_files_selection(
+    data_dir: Path, has_download_dir: bool, files_expected: dict, expectation: ContextManager
+) -> None:
+    """Test the get a subset of download.get_files_selection() files function `
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:336]:
```suggestion
    """Tests the `download.get_files_selection()` function.
```

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]
A great update of the module and nice handling of the unit tests. Most of the comments are more on docstring formatting than anything else.

A couple of details:
* `@pytest.mark.dependency()` should only be added to those methods that depend on others or other tests will depend on them (have flagged a couple of cases but this should be reviewed and updated).
* I'm removing the reference to `_FILE_ENDS` in docstrings because it is a private constant, so this information should not be exposed. If it is relevant maybe it should become public.
* Consider a rename of `dl_dir` arguments/variables to `download_dir`, as I do not find the former too readable.

File: src/python/ensembl/io/genomio/assembly/download.py [hunk 18]
```diff
@@ -193,10 +240,16 @@ def _download_file(
 
 def get_files_selection(dl_dir: Path) -> Dict[str, str]:
     """
-    Among all the files downloaded, only keep a subset for which we use a controlled name.
-    Return a dict[name] = file_path
-    The file_path is relative to the download dir
-    Current names are defined in _FILE_ENDS
+    """Returns a dictionary with the relevant downloaded files classified.
```
Comment (JAlvarezJarreta) [inline src/python/ensembl/io/genomio/assembly/download.py:243]: Forgot to remove the `"""` from the previous line.

File: src/python/tests/assembly/test_download.py [hunk 19]
```diff
@@ -0,0 +1,458 @@
...
+            Path("malformed_md5_checksums.txt"),
+            None,
+            pytest.raises(ValueError),
+            id="Bad md5 file - tab separated",
+        ),
+        pytest.param(Path("dir_name"), None, does_not_raise(), id="Dir passed instead of file"),
+    ],
+)
+def test_checksums(
+    data_dir: Path, checksum_file: Optional[Path], checksum: str, expectation: ContextManager
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:109]:
```suggestion
    data_dir: Path, checksum_file: Path, checksum: Optional[str], expectation: ContextManager
```
I believe I messed up the `Optional` here, sorry! (highlighted by `mypy`)

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]
Just a couple of minor issues that are causing most CI/CD checks to fail.

File: src/python/tests/assembly/test_download.py [hunk 20]
```diff
@@ -0,0 +1,458 @@
...
+        ),
+        pytest.param(
+            "non-existing-file.txt",
+            {"non-existing-file.txt": "e98b980b442fdb2a21877dcc55e11848"},
+            pytest.raises(ftp_error_reply),
+            id="File has md5sum but does not exist",
+        ),
+        pytest.param(
+            "test_ftp_file.txt",
+            {"test_ftp_file.txt": ""},
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:177]:
```suggestion
            {"test_ftp_file.txt": "e98b980b442fdb2a21877dcc55e11848"},
```
Missing checksum to force retry.

File: src/python/tests/assembly/test_download.py [hunk 21]
```diff
@@ -0,0 +1,458 @@
...
+            pytest.raises(FileDownloadError),
+            id="Incorrect md5sum",
+        ),
+    ],
+)
+@patch("ftplib.FTP")
+def test_download_single_file(
+    mock_ftp: Mock,
+    data_dir: Path,
+    tmp_dir: Path,
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:193]:
```suggestion
    tmp_path: Path,
```

File: src/python/tests/assembly/test_download.py [hunk 22]
```diff
@@ -0,0 +1,458 @@
...
+    ftp_file: str,
+    md5_sums: dict,
+    expectation: ContextManager,
+) -> None:
+    """Tests the private function _download_file.
+
+    Args:
+        mock_ftp: Mock FTP object
+        data_dir: Path to test data root dir
+        tmp_dir: Temp dir created for test
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:203]:
```suggestion
        tmp_path: Temp dir created for test
```

File: src/python/tests/assembly/test_download.py [hunk 23]
```diff
@@ -0,0 +1,458 @@
...
+        ftp_file: FTP file which to mock download
+        md5_sums: FTP file and md5_sum value pair
+        expectation: Context manager expected raise exception
+    """
+
+    data_file = data_dir / ftp_file
+    retr_file = tmp_dir / ftp_file
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:210]:
```suggestion
    retr_file = tmp_path / ftp_file
```

File: src/python/tests/assembly/test_download.py [hunk 24]
```diff
@@ -0,0 +1,458 @@
...
+        try:
+            with data_file.open("rb") as data_fh:
+                callback(data_fh.read())
+        except OSError as err:
+            raise ftp_error_reply from err
+
+    mock_ftp.retrbinary.side_effect = mock_retr_binary
+
+    with expectation:
+        _download_file(mock_ftp, ftp_file, md5_sums, tmp_dir)
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:223]:
```suggestion
        _download_file(mock_ftp, ftp_file, md5_sums, tmp_path)
```

File: src/python/ensembl/io/genomio/assembly/download.py (hunk 18 above)
Comment (ens-LCampbell) [reply src/python/ensembl/io/genomio/assembly/download.py:242]: Removed the extra triple quote thanks @JAlvarezJarreta

**Review by MatBarba — CHANGES_REQUESTED** [review]
When I run pytest the testing hangs for a few seconds for the test `test_retrieve_assembly_data[Case 3: Download files list empty]`. I suspect that it might be connecting to something, so something might not be mocked completely for this case. Can you check?

File: src/python/tests/assembly/test_download.py [hunk 25]
```diff
@@ -0,0 +1,458 @@
...
+            "GCA_017607445.1",
+            True,
+            0,
+            False,
+            pytest.raises(FileDownloadError),
+            id="Case 3: Download files list empty",
+        ),
+    ],
+)
+@patch("ftplib.FTP")
```
Comment (MatBarba) [inline src/python/tests/assembly/test_download.py:400]: You need to patch the FTP object that you use in the module (autospec ensures all its methods are also mocked):
```suggestion
@patch("ensembl.io.genomio.assembly.download.FTP", autospec=True)
```

**Review by MatBarba — CHANGES_REQUESTED** [review]
The cause of the hang I mentioned is due to the location where FTP is patched:

This patches FTP in the test module, but not the FTP used in the download module:
```
@patch("ftplib.FTP")
```

This fixes it:
```
@patch("ensembl.io.genomio.assembly.download.FTP", autospec=True)
```
This should be applied to all tests that mock FTP

File: src/python/tests/assembly/test_download.py [hunk 26]
```diff
@@ -0,0 +1,458 @@
...
+        ),
+        pytest.param(
+            "ftp.ncbi.nlm.nih.gov",
+            "gcx_017607445.1",
+            pytest.raises(UnsupportedFormatError),
+            id="Failed connection improper INSDC accession",
+        ),
+    ],
+)
+@patch("ftplib.FTP")
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:62]:
```suggestion
@patch("ensembl.io.genomio.assembly.download.FTP", autospec=True)
```

File: src/python/tests/assembly/test_download.py [hunk 27]
```diff
@@ -0,0 +1,458 @@
...
+        ),
+        pytest.param(
+            "test_ftp_file.txt",
+            {"test_ftp_file.txt": "e98b980b442fdb2a21877XxxXxxXxxXx"},
+            pytest.raises(FileDownloadError),
+            id="Incorrect md5sum",
+        ),
+    ],
+)
+@patch("ftplib.FTP")
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:189]:
```suggestion
@patch("ensembl.io.genomio.assembly.download.FTP", autospec=True)
```

File: src/python/tests/assembly/test_download.py [hunk 28]
```diff
@@ -0,0 +1,458 @@
...
+            "GCA_017607445.1",
+            "GCA_544706710.1",
+            "md5checksums_ftp_source.txt",
+            does_not_raise(),
+            2,
+            id="Error case: GCA doesn't match ftp_dir path",
+        ),
+    ],
+)
+@patch("ftplib.FTP")
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:252]:
```suggestion
@patch("ensembl.io.genomio.assembly.download.FTP", autospec=True)
```

File: src/python/tests/assembly/test_download.py [hunk 29]
```diff
@@ -0,0 +1,458 @@
...
+    ftp_accession: str,
+    compare_accession: str,
+    md5: str,
+    exception: ContextManager,
+    max_redo: int,
+) -> None:
+    """Tests the download.download_files() function
+
+    Args:
+        mock_ftp: Magic Mock of ftplib.FTP object
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:266]:
```suggestion
        mock_ftp: Mock of `ensembl.io.genomio.assembly.download.FTP` object
```

**Review by JAlvarezJarreta — APPROVED** [review]
Provided @MatBarba 's suggestions.

File: src/python/tests/assembly/test_download.py [hunk 30]
```diff
@@ -0,0 +1,458 @@
...
+        mock_ftp.cwd.assert_called_with("genomes/all/GCA/017/607/445")
+
+        download_files(connected_ftp, compare_accession, data_dir, max_redo)
+        mock_ftp.cwd.assert_called()
+        mock_ftp.mlsd.assert_called()
+
+        if ftp_accession == compare_accession:
+            mock_ftp.retrbinary.assert_called()
+        elif ftp_accession != compare_accession:
+            mock_ftp.retrbinary.not_called()
```
Comment (JAlvarezJarreta) [inline src/python/tests/assembly/test_download.py:311]:
```suggestion
        else:
            mock_ftp.retrbinary.assert_not_called()
```

**Pull request merged**